torch/csrc/jit/passes/graph_fuser.cpp

```diff
@@ -244,6 +275,58 @@ struct GraphFuser {
     return *n->g(attr::Subgraph);
   }
 
+  void decomposeLayerNorm(Node* layer_norm) {
+    static std::shared_ptr<Graph> ln_graph;
+    static std::once_flag flag;
+    std::call_once(
+        flag,
+        [](std::shared_ptr<Graph>* graph_ptr) {
+          static const char* source = R"SCRIPT(
+        def layer_norm(input : Tensor, normalized_shape : List[int], eps : float, cudnn_enable : bool) -> Tensor:
+            input_ndim = input.dim()
+            normalized_ndim = len(normalized_shape)
+            n = 1
+            for i in range(input_ndim - normalized_ndim):
+                n *= input.size(i)
+            input_reshape = input.contiguous().view(1, n, -1)
+            mean, invstd = torch.batch_norm_stats(input_reshape, eps)
+            input_shape = input.size()
+            mean = torch._ncf_view(mean, input_shape, normalized_ndim)
+            invstd = torch._ncf_view(invstd, input_shape, normalized_ndim)
+
+            return (input - mean) * invstd
+      )SCRIPT";
+          auto module = std::make_shared<script::Module>();
+          defineMethodsInModule(
+              module, source, script::nativeResolver, /*self=*/c10::nullopt);
+          *graph_ptr = module->get_method("layer_norm").graph();
+        },
+        &ln_graph);
+
+    AT_ASSERT(isFusableLayerNorm(layer_norm));
+    WithInsertPoint insert_guard{layer_norm};
+    Value* input = layer_norm->namedInput(attr::input);
+    std::vector<Value*> inputs{input,
+                               layer_norm->namedInput(attr::normalized_shape),
+                               layer_norm->namedInput(attr::eps),
+                               layer_norm->namedInput(attr::cudnn_enable)};
+    Value* new_output =
+        SubgraphUtils::inlineGraph(ln_graph, inputs, layer_norm).at(0);
+    auto weight = layer_norm->namedInput(attr::weight);
+    auto bias = layer_norm->namedInput(attr::bias);
+    auto weight_defined = isDefined(weight).value();
+    auto bias_defined = isDefined(bias).value();
+    if (weight_defined && bias_defined) {
+      new_output = graph_->insert(aten::addcmul, {bias, new_output, weight});
+    } else if (weight_defined) {
+      new_output = graph_->insert(aten::mul, {new_output, weight});
+    } else if (bias_defined) {
+      new_output = graph_->insert(aten::add, {new_output, bias});
+    }
+    layer_norm->output()->replaceAllUsesWith(new_output);
+    layer_norm->destroy();
```

Comment:
This infra should also be shared. THey take almost the same arguments, only need a different reshape and replacement.